torch/nn/parallel/distributed.py

```diff
@@ -108,6 +109,19 @@ class _DDPUnevenInputsConfig(NamedTuple):
     ddp_join_enabled: bool
     ddp_join_divide_by_initial_world_size: bool
 
+# Add a DDPSink to queue call back of out-most backward/graph task,
+# this helps call back is fired after all gradients' calculation
+# is completed.
+class DDPSink(Function):
+    @staticmethod
+    def forward(ctx, input, reducer):
+        ctx.reducer = reducer
+        return input
+
+    @staticmethod
+    def backward(ctx, input):
+        Variable._execution_engine.queue_callback(ctx.reducer._delay_all_reduce)
```

Comment:
Why do we need a DDPSink function like this? Can't all the delay_allreduce logic go into `finalize_backward`?

Reply:
right now finalize_backward is queued when all gradients are completed. 

1) at first iteration, we do not know when the gradients is completed, so we can not queue in the same way as how finalize_backward is queued. 2) we can queue the callback before backward starts, but it needs to be associated with the out most backward task

Reply:
I think the current version of this PR works well, but as an enhancement, we could consider:

1) Don't wait to enqueue finalize_backward until all buckets have allreduce launched, instead, enqueue it immediately and refactor it to block waiting for allreduces to be kicked off (an idea @pritamdamania87 mentioned)
2) Move enqueueing to this sink function
3) For static graph, we'd then queue either delay allreduce or regular finalize_backward depending on the business logic.

This has the benefit that there's only a single well-defined place where the callback is being enqueued.

Reply:
move queueing finalize_backward to backward end should work, but we should be careful about it. e..g, not forgot to checking all buckets are computed and etc